sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +212,248 @@ object Hash {
       val ss = set.toSeq.sorted[String]
       iterateOver(ss.iterator, ss.size)(_ addString _)
     }
+
+    final def addOptional[S](opt: Option[S], hashS: this.type => S => this.type): this.type = {
+      opt match {
+        case None => addInt(0)
+        case Some(value) => hashS(addInt(1))(value)
+      }
+    }
+  }
+
+  // Transaction Builder class with builder methods to recursively encode nodes
+  private sealed class TransactionBuilder(
+      protected val messageDigest: MessageDigestWithContext,
+      valueBuilder: ValueHashBuilder,
+  ) extends Builder(stringNumericToBytes, messageDigest.isContextAware) {
```

Comment:
Out of curiosity: Why so we serialize numerics via strings for the transaction builder, but use a byte array encoding otherwise?

Reply:
The string is easier to specified and implement by an client (not implement by DA).
We keep the old one for backward compatibility.